### ISSUE TITLE
因为0**0=1，所以需要增加γ=0时的特殊情况。否则会出现问题

### ISSUE THREAD

**Issue by zhangyuqi-1** [ailias/Focal-Loss-implement-on-Tensorflow#14]
```
#因为0**0=1，所以需要增加γ=0时的特殊情况

def focal_loss_sigmoid(prediction_tensor, target_tensor, weights=None, alpha=0.5, gamma=0):
    target_tensor = tf.cast(target_tensor, tf.float32)
    sigmoid_p = tf.nn.sigmoid(prediction_tensor)
    zeros = array_ops.zeros_like(sigmoid_p, dtype=sigmoid_p.dtype)
    pos_p_sub = array_ops.where(target_tensor > zeros, target_tensor - sigmoid_p, zeros)
    neg_p_sub = array_ops.where(target_tensor > zeros, zeros, sigmoid_p)
    if gamma != 0:
        per_entry_cross_ent = - alpha * (pos_p_sub ** gamma) * tf.log(tf.clip_by_value(sigmoid_p, 1e-8, 1.0)) \
                              - (1 - alpha) * (neg_p_sub ** gamma) * tf.log( tf.clip_by_value(1.0 - sigmoid_p, 1e-8, 1.0))
    else:
        # ∵0**0=1 
        per_entry_cross_ent = - alpha * target_tensor * tf.log(tf.clip_by_value(sigmoid_p, 1e-8, 1.0)) \
                              - (1 - alpha) * (1 - target_tensor) * tf.log(tf.clip_by_value(1.0 - sigmoid_p, 1e-8, 1.0))
    return per_entry_cross_ent
```

**Comment by zhangyuqi-1** [comment]
> def focal_loss_sigmoid(prediction_tensor, target_tensor, weights=None, alpha=0.5, gamma=0):
> target_tensor = tf.cast(target_tensor, tf.float32)
> sigmoid_p = tf.nn.sigmoid(prediction_tensor)
> zeros = array_ops.zeros_like(sigmoid_p, dtype=sigmoid_p.dtype)
> 
> ```
> # For poitive prediction, only need consider front part loss, back part is 0;
> # target_tensor > zeros <=> z=1, so poitive coefficient = z - p.
> pos_p_sub = array_ops.where(target_tensor > zeros, target_tensor - sigmoid_p, zeros)
> 
> # For negative prediction, only need consider back part loss, front part is 0;
> # target_tensor > zeros <=> z=1, so negative coefficient = 0.
> neg_p_sub = array_ops.where(target_tensor > zeros, zeros, sigmoid_p)
> 
> if gamma != 0:
>     per_entry_cross_ent = - alpha * (pos_p_sub ** gamma) * tf.log(tf.clip_by_value(sigmoid_p, 1e-8, 1.0)) \
>                           - (1 - alpha) * (neg_p_sub ** gamma) * tf.log(
>         tf.clip_by_value(1.0 - sigmoid_p, 1e-8, 1.0))
> else:
>     # ∵0**0=1 
>     per_entry_cross_ent = - alpha * target_tensor * tf.log(tf.clip_by_value(sigmoid_p, 1e-8, 1.0)) \
>                           - (1 - alpha) * (1 - target_tensor) * tf.log(tf.clip_by_value(1.0 - sigmoid_p, 1e-8, 1.0))
> 
> return per_entry_cross_ent
> ```
